Create KFServing Python SDK

jinchihe opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://www.kubeflow.org/docs/about/contributing/ and developer guide https://github.com/kubeflow/kfserving/blob/master/docs/DEVELOPER_GUIDE.md
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
The PR is to create KFServing Python SDK, according to instructions from @yuzisun in the #160, and you can see the practice in the `hack/sdk-gen/README.md` in the PR.

**Which issue(s) this PR fixes** :
Fixes #160

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Created KFServing Python SDK
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/kfserving/218)
<!-- Reviewable:end -->

jinchihe commented:
/assign @yuzisun 

/retest

ellis-bigelow commented on hack/sdk-gen/README.md:
````diff
@@ -0,0 +1,35 @@
+go install -v
+```
+
+### Generate the openapi model from kersering types
````
kersering

jinchihe commented:
> Hey this is pretty cool. Do you mind separating the CLs from generated code and from the code that generates it. It will make this easier to review.

@ellis-bigelow the `openapi-gen` (already in vendor) and `swagger-codegen` is ignored before checkin, in fact all commits are generated files. And do you think (or mean) to separate openapi model (openapi-gen generated) and  and python sdk (swagger-codegen generated)? If so, we may can do that like TFJob done in [update-codegen.sh](https://github.com/kubeflow/tf-operator/blob/e1996e2407e0bea594b0db4fb24c49e4617799a1/hack/update-codegen.sh#L65).

jinchihe commented:
/hold

yuzisun commented on hack/sdk-gen/main.go:
```diff
@@ -0,0 +1,78 @@
+limitations under the License.
+*/
+
+package main
```
I think we can move this to cmd directory, also this binary is generating the swagger.json so maybe name it `spec-gen`

jinchihe replied:
Great idea! Thanks.

jinchihe commented:
Added some input-dirs for openapi-gen, otherwise, there will be import error during tests. Thanks.

yuzisun commented on hack/sdk-gen/README.md:
````diff
@@ -0,0 +1,46 @@
+
+Update the following kative files to add `+k8s:openapi-gen=true` (This should be updated from knative).
+```
+vendor/github.com/knative/pkg/apis/condition_types.go
````
hmm not sure if we should change files in vendor, next person who run `dep ensure` can overwrite all these changes.

jinchihe replied:
Yes, your point is absolutely correct! So I mentioned this in README file so that following contributor can do this manually before running the command, and the changes are not committed. In fact the best way to enhance this from knative side (may send a PR to update this later for knative). For now, we can just update that manually otherwise the knative part will be skipped taht will cause import errors.

jinchihe replied:
Send a [PR](https://github.com/knative/pkg/pull/510) to update in knative pkg, but now I think we need to update that before executing openapi-ge.

yuzisun commented on hack/sdk-gen/README.md:
````diff
@@ -0,0 +1,46 @@
+### Generate swagger file
+
+```
+go run hack/sdk-gen/main.go 0.1 > hack/sdk-gen/swagger.json
````
Add this command to `Make codegen` ?

jinchihe replied:
OK, I will add into `update-codegen.sh` which will be called in Makefile, for update knative files, generating OpenAPI specification and generating swagger file.

jinchihe replied:
Reexamined, I think that's better to create separated script for the openapi-gen from update-codegen.sh, so added one script `update-openapigen.sh` in the patch, and added into the Makefile.

yuzisun commented on hack/sdk-gen/README.md:
````diff
@@ -0,0 +1,46 @@
+```
+go run hack/sdk-gen/main.go 0.1 > hack/sdk-gen/swagger.json
+```
+Note the version need to be supplied for the `main.go`, need to update the release version for future release. After the executing, the swagger generated to `hack/sdk-gen/swagger.json`.
````
`swagger.json` should be colocated with `openapi_generated.go` ?

jinchihe replied:
OK, will put to some location with `openapi_generated.go`.

yuzisun commented on hack/sdk-gen/README.md:
```diff
@@ -0,0 +1,46 @@
+
+### Install swagger-codegen
+
+Following the [guide](https://github.com/swagger-api/swagger-codegen#getting-started) to install swagger-codegen, the following step assumps swagger-codegen has been installed under `hack` directory.
```
/assumps/assumes/

yuzisun commented on hack/sdk-gen/README.md:
```diff
@@ -0,0 +1,46 @@
+
+Following the [guide](https://github.com/swagger-api/swagger-codegen#getting-started) to install swagger-codegen, the following step assumps swagger-codegen has been installed under `hack` directory.
+
+### Generate Python SDK from the swagger file
```
maybe we can write a simple script which just download the `swagger-codegen-cli.jar` and generate sdk so that contributor does not have to take extra step to figure out swagger stuff ?

jinchihe replied:
OK, will draft a simple script to download jar and run generating.

yuzisun commented on sdk/docs/ComGithubKnativePkgApisCondition.md:
```diff
@@ -0,0 +1,15 @@
+# ComGithubKnativePkgApisCondition
```
Is there way to remove `ComGithub`?

jinchihe replied:
Yes, we should remove the `ComGithub`.

yuzisun commented:
Awesome work!  Looks the core kubernetes python object model are already here  https://github.com/kubernetes-client/python/tree/master/kubernetes/client/models, I guess we can just add this dependency to our sdk?

yuzisun commented:
> > Hey this is pretty cool. Do you mind separating the CLs from generated code and from the code that generates it. It will make this easier to review.
> 
> @ellis-bigelow the `openapi-gen` (already in vendor) and `swagger-codegen` is ignored before checkin, in fact all commits are generated files. And do you think (or mean) to separate openapi model (openapi-gen generated) and and python sdk (swagger-codegen generated)? If so, we may can do that like TFJob done in [update-codegen.sh](https://github.com/kubeflow/tf-operator/blob/e1996e2407e0bea594b0db4fb24c49e4617799a1/hack/update-codegen.sh#L65).

@ellis-bigelow I am quite familiar with all the generations, it is ok to leave as is to review this time :).

jinchihe commented:
> Awesome work! Looks the core kubernetes python object model are already here https://github.com/kubernetes-client/python/tree/master/kubernetes/client/models, I guess we can just add this dependency to our sdk?

@yuzisun It's really appreciated for your great comments!  For the dependency, yes, kubernetes python clients is exists , but not sure we can depends on that, but will have more investigation for this. Thanks.

jinchihe commented:
@yuzisun Thanks for your mentoring. I updated the code changed, removed the k8s existing stuff, and used the "importMappings"  to set dependency from k8s in swagger_config.json file. 

Here it's worth mentioning that, for `io_k8s_apimachinery_pkg_apis_meta_v1_time.py` that's called by knative, I did not get that from official k8s python client repo, so added in this PR.

Sample demo for the SDK here.
```
(python36) [root@jinchi1 sdk]# cat mytest.py
from kubernetes import client

from kfserving.models.kfserving_v1alpha1_model_spec import KfservingV1alpha1ModelSpec as V1alpha1ModelSpec
from kfserving.models.kfserving_v1alpha1_tensorflow_spec import KfservingV1alpha1TensorflowSpec as V1alpha1TensorflowSpec
from kfserving.models.kfserving_v1alpha1_kf_service import KfservingV1alpha1KFService as V1alpha1KFService
from kfserving.models.kfserving_v1alpha1_kf_service_spec import KfservingV1alpha1KFServiceSpec as V1alpha1KFServiceSpec

model_spec =  V1alpha1ModelSpec(tensorflow=V1alpha1TensorflowSpec(model_uri='gs://flower'))
kfsvc = V1alpha1KFService(metadata=client.V1ObjectMeta(name='flower'),spec= V1alpha1KFServiceSpec(default=model_spec))

print(kfsvc)
```
Executing the sample python script:
```
(python36) [root@jinchi1 sdk]# python mytest.py
{'api_version': None,
 'kind': None,
 'metadata': {'annotations': None,
              'cluster_name': None,
              'creation_timestamp': None,
              'deletion_grace_period_seconds': None,
              'deletion_timestamp': None,
              'finalizers': None,
              'generate_name': None,
              'generation': None,
              'initializers': None,
              'labels': None,
              'name': 'flower',
              'namespace': None,
              'owner_references': None,
              'resource_version': None,
              'self_link': None,
              'uid': None},
 'spec': {'canary': None,
          'canary_traffic_percent': None,
          'default': {'custom': None,
                      'max_replicas': None,
                      'min_replicas': None,
                      'pytorch': None,
                      'service_account_name': None,
                      'sklearn': None,
                      'tensorflow': {'model_uri': 'gs://flower',
                                     'resources': None,
                                     'runtime_version': None},
                      'tensorrt': None,
                      'xgboost': None}},
 'status': None}
(python36) [root@jinchi1 sdk]#
```

jinchihe commented:
/hold cancel

yuzisun commented on sdk/test/test_kfserving_v1alpha1_custom_spec.py:
```diff
@@ -0,0 +1,40 @@
+    def tearDown(self):
+        pass
+
+    def testKfservingV1alpha1CustomSpec(self):
```
not sure what these tests are supposed to do since we are just going to call `CustomObjectApi` to create KFService.

jinchihe replied:
I suggest keeping that since that's generated by swagger-codegen, at least the case can test the import/load has no problem. Anyway, we need to make and clear up the test cases after we developed some action apis :-）

ellis-bigelow commented on hack/sdk-gen/README.md:
```diff
@@ -0,0 +1,25 @@
+
+## Generate openapin spec and swagger file. 
+
+From KFSering root folder, run the below script to generate openapin spec and swagger file.
```
KFServing typo

ellis-bigelow replied:
openapin typo.

jinchihe replied:
Updated the typo. Thanks!

jinchihe commented on hack/sdk-gen/sdk-gen.sh:
```diff
@@ -0,0 +1,29 @@
+wget -O ${SWAGGER_CODEGEN_JAR} ${SWAGGER_JAR_URL} >> /dev/null 2>&1
+if [ $? -ne 0 ]; then
+    echo "ERROR: Failed to download swagger-codege jar pacakge."
+    exit 1
```
Updated the script. Thanks.

jinchihe commented on hack/sdk-gen/swagger_config.json:
```diff
@@ -0,0 +1,11 @@
+        "IoK8sApiCoreV1Container": "from kubernetes.client import V1Container as IoK8sApiCoreV1Container",
+        "IoK8sApimachineryPkgApisMetaV1ObjectMeta": "from kubernetes.client import V1ObjectMeta as IoK8sApimachineryPkgApisMetaV1ObjectMeta",
+        "IoK8sApimachineryPkgApisMetaV1ListMeta": "from kubernetes.client import V1ListMeta as IoK8sApimachineryPkgApisMetaV1ListMeta",
+        "IoK8sApiCoreV1ResourceRequirements": "from kubernetes.client import V1ResourceRequirements as IoK8sApiCoreV1ResourceRequirements"
```
Enhanced as your comments. Thanks.

jinchihe commented on sdk/docs/KfservingV1alpha1KFServiceList.md:
```diff
@@ -0,0 +1,13 @@
+**api_version** | **str** | APIVersion defines the versioned schema of this representation of an object. Servers should convert recognized schemas to the latest internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#resources | [optional] 
+**items** | [**list[KfservingV1alpha1KFService]**](KfservingV1alpha1KFService.md) |  | 
+**kind** | **str** | Kind is a string value representing the REST resource this object represents. Servers may infer this from the endpoint the client submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#types-kinds | [optional] 
+**metadata** | [**IoK8sApimachineryPkgApisMetaV1ListMeta**](IoK8sApimachineryPkgApisMetaV1ListMeta.md) |  | [optional] 
```
Updated the doc file manually. Thanks!

jinchihe commented on sdk/kfserving/models/kfserving_v1alpha1_sk_learn_spec.py:
```diff
@@ -0,0 +1,174 @@
+from kubernetes.client import V1ResourceRequirements as IoK8sApiCoreV1ResourceRequirements  # noqa: F401,E501
+
+
+class KfservingV1alpha1SKLearnSpec(object):
```
Dropped the kfserving from class name. Thanks :-)

jinchihe commented on hack/update-openapigen.sh:
```diff
@@ -0,0 +1,38 @@
+go run vendor/k8s.io/code-generator/cmd/openapi-gen/main.go --input-dirs github.com/kubeflow/kfserving/pkg/apis/serving/v1alpha1,github.com/knative/pkg/apis --output-package github.com/kubeflow/kfserving/pkg/apis/serving/v1alpha1/ --go-header-file hack/boilerplate.go.txt
+
+# Workaroud error "spec redeclared in this block" in unit test.
+sed -i 's%spec "github.com/go-openapi/spec"%openapispec "github.com/go-openapi/spec"%g' pkg/apis/serving/v1alpha1/openapi_generated.go
```
Yes, we can, but I suggest keeping this, since not only for unit test, but also for usual cases. the `spec` has exists in the [pkg/apis/serving/v1alpha1/kfservice_framework_tensorrt_test.go](https://github.com/kubeflow/kfserving/blob/9b91e1a6030ebf75d9ee33946405d2e42dd803ea/pkg/apis/serving/v1alpha1/kfservice_framework_tensorrt_test.go#L41), we should have alias for spec, otherwise there is error if run `go vet`. And even if no `spec` existing, we should avoid to use `spec` as it's too common. so I suggest to update that after generated. Thanks.

jinchihe commented on hack/update-codegen.sh:
```diff
@@ -24,4 +24,19 @@ if [ -z "${GOPATH:-}" ]; then
+
+# Generating OpenAPI specification
+go run vendor/k8s.io/code-generator/cmd/openapi-gen/main.go --input-dirs github.com/kubeflow/kfserving/pkg/apis/serving/v1alpha1,github.com/knative/pkg/apis/duck/v1beta1,github.com/knative/pkg/apis,k8s.io/api/core/v1,k8s.io/apimachinery/pkg/apis/meta/v1,k8s.io/apimachinery/pkg/api/resource,k8s.io/apimachinery/pkg/runtime,k8s.io/apimachinery/pkg/util/intstr,k8s.io/apimachinery/pkg/version  --outpu ...
+# Workaroud error "spec redeclared in this block" in unit test.
```
Make resolve the conversation since same with below. Trace in below conversation. Thanks.

jinchihe commented:
Seems the kubeflow-kfserving-presubmit testing env has problem, will retest once the env ready. Thanks.

yuzisun commented:
Great work! This looks pretty good to me now, a few final comments.

yuzisun commented:
@ellis-bigelow Do you have more comments on this?

ellis-bigelow commented:
I'm good if you are.

yuzisun commented:
/lgtm
/approve

jinchihe commented:
/retest
